webapp/app/[locale]/tunnel/_components/withdraw.tsx

```diff
@@ -50,19 +69,132 @@ type BtcWithdrawProps = {
-  const { fromToken, updateFromInput } = state
+  // use this to hold the withdrawn amount for the Tx list after clearing the state upon confirmation
+  const [amountWithdrawn, setAmountWithdrawn] = useState('0')
+  const [isWithdrawing, setIsWithdrawing] = useState(false)
+
+  const {
+    fromInput,
+    fromNetworkId,
+    fromToken,
+    resetStateAfterOperation,
+    updateFromInput,
+  } = state
+
+  const { btcAddress, evmChainId } = useAccounts()
+  const fromChain = useChain(fromNetworkId)
+  const { balance: bitcoinBalance } = useTokenBalance(fromToken)
   const t = useTranslations()
-  const isWithdrawing = false
-  // eslint-disable-next-line arrow-body-style
-  const handleWithdraw = () => {}
+  const { updateWithdrawal, withdrawals } = useTunnelHistory()
+  const {
+    clearWithdrawBitcoinState,
+    withdrawBitcoin,
+    withdrawBitcoinReceipt,
+    withdrawBitcoinReceiptError,
+    withdrawError,
+    withdrawTxHash,
+  } = useWithdrawBitcoin()
+
+  const resetFormState = useCallback(
+    function () {
+      setIsWithdrawing(false)
+      resetStateAfterOperation()
+    },
+    [setIsWithdrawing, resetStateAfterOperation],
+  )
+
+  const onSuccess = useCallback(
+    function () {
+      resetFormState()
+
+      const withdrawalFound = withdrawals.find(
+        w =>
+          w.transactionHash === withdrawBitcoinReceipt.transactionHash &&
+          !w.timestamp,
+      )
+
+      const extendedWithdrawal = {
+        ...withdrawalFound,
+        blockNumber: Number(withdrawBitcoinReceipt.blockNumber),
+      }
+
+      // Handling of this error is needed https://github.com/hemilabs/ui-monorepo/issues/322
+      // eslint-disable-next-line promise/catch-or-return
+      addTimestampToOperation(extendedWithdrawal, fromNetworkId).then(
+        ({ timestamp }) =>
+          updateWithdrawal(extendedWithdrawal, {
+            status: MessageStatus.RELAYED,
+            timestamp,
+          }),
+      )
+    },
+    [
+      fromNetworkId,
+      resetFormState,
+      updateWithdrawal,
+      withdrawals,
+      withdrawBitcoinReceipt,
+    ],
+  )
+
+  const { beforeTransaction } = useAfterTransaction({
+    clearState: clearWithdrawBitcoinState,
+    errorReceipts: [withdrawError, withdrawBitcoinReceiptError],
+    onError: resetFormState,
+    onSuccess,
+    transactionReceipt: withdrawBitcoinReceipt,
+  })
+
+  const handleWithdraw = function () {
+    beforeTransaction()
+    setAmountWithdrawn(fromInput)
+    clearWithdrawBitcoinState()
+    withdrawBitcoin(parseUnits(fromInput, fromToken.decimals))
+    setIsWithdrawing(true)
+  }
+
+  // TODO we need to let the user know about the min value to withdraw
+  // See https://github.com/hemilabs/ui-monorepo/issues/454
+  const canWithdraw =
+    canSubmit({
+      balance: bitcoinBalance,
+      chainId: evmChainId,
+      fromInput,
+      fromNetworkId,
+      fromToken,
+    }) && Big(fromInput).gte(MinBitcoinWithdraw)
+
+  const gas = {
+    // TODO GET gas estimation
```

Comment:
Is this OK?

Reply:
No, it is wrong! Here I added the estimation [d188b3d](https://github.com/hemilabs/ui-monorepo/pull/461/commits/d188b3dd5b35de2742cdaced11e75e5446edf287)

I haven't tested it because the testnet's fork is not working 😢  but it should work.